### PR TITLE
Improve `fld` documentation

### LINKED ISSUE #17415: fld(1,0.2) returns 4.0
I'm totally not sure why this happens. Does anyone know if this is spec or bug?

``` jl
julia> 1/0.2
5.0

julia> floor(1/0.2)
5.0

julia> fld(1,0.2)
4.0
```

I'm using Mac OSX and Julia version is `0.4.6`.

### PR DESCRIPTION
**Pull request by ranjanan**
This changes the docstring to "Largest integer n such that y \* n less than or equal to x", at @yuyichao 's suggestion, and general approval in #17415.

Also, this fixes #17415.

### CODE REVIEW

**Comment by yuyichao** [comment]
Better to just force push next time. Maybe still better to mention the case in https://github.com/JuliaLang/julia/issues/17415 somewhere in the doc. The new doc is probably accurate but can still be confusing when people hit those cases. Just mentioning that this can be different from `floor(x / y)` due to floating point rounding is probably good enough.

File: base/docs/helpdb/Base.jl [hunk 1]
```diff
@@ -2603,7 +2603,8 @@ creates a task, and does not run it.
 """
     fld(x, y)
 
-Largest integer less than or equal to `x/y`.
+Largest integer `n` such that `y * n` less than or equal to `x`. Note that this can be
+different from `floor(x/y)` due to floating point rounding.
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: could you run `make docs` and commit the generated rst? thanks!

File: base/docs/helpdb/Base.jl [hunk 2]
```diff
@@ -2603,7 +2603,8 @@ creates a task, and does not run it.
 """
     fld(x, y)
 
-Largest integer less than or equal to `x/y`.
+Largest integer `n` such that `y * n` less than or equal to `x`. Note that this can be
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: such that `y * n` is less than ...

**Comment by ranjanan** [comment]
@tkelman Done.

File: base/docs/helpdb/Base.jl [hunk 3]
```diff
@@ -2603,7 +2603,8 @@ creates a task, and does not run it.
 """
     fld(x, y)
 
-Largest integer less than or equal to `x/y`.
+Largest integer `n` such that `y * n` is less than or equal to `x`. Note that this can be
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: So is this the current definition, or was @yuyichao saying we'd need to change the code to make the implementation match this behavior in https://github.com/JuliaLang/julia/issues/17415#issuecomment-232768152?

Comment (ranjanan) [inline base/docs/helpdb/Base.jl]: I am under the impression that @yuyichao was saying (in that comment) that the documentation should change to reflect the behavior of the function. So, I changed the docs appropriately.

Comment (tkelman) [inline base/docs/helpdb/Base.jl]: depends whether

> the definition of `fld`

was referring to definition as in behavior of the code, or what we document. We'd like them to be the same, but which is the code doing right now?

Comment (eschnett) [inline base/docs/helpdb/Base.jl]: The generic definition is
```Julia
fld{T<:Real}(x::T,y::T) = convert(T,round((x-mod(x,y))/y))
```
where `mod` could presumably be implemented as
```
mod{T<:Real}(x::T,y::T) = x-y*floor(x/y)
```
If you remove some extraneous gymnastics (multiply by / divide by `y`, subtract from `x` twice) and ignore the associated rounding errors, this leads to
```Julia
fld(x,y) = floor(x/y)
```
which is probably also what most people expect as it is direct and simple. "fld" means "flooring division", so this is the definition I'd suggest to use. (And then we can update and simplify the code, and add tests for cases where round-off errors kick in.)

Comment (yuyichao) [inline base/docs/helpdb/Base.jl]: From https://github.com/JuliaLang/julia/issues/17415#issuecomment-232730858 it seems that the current implementation is what it is meant to be. The original doc was also not wrong, just misleading.

**Comment by ranjanan** [comment]
@tkelman Is there anything holding up this issue? Any way I can make the PR better?

**Comment by tkelman** [comment]
I still didn't get a clear answer out of those comments on which it should be.

**Comment by eschnett** [comment]
I think the current description "Largest integer less than or equal to `x/y`" is correct. This is also how e.g. `floor` is documented. Whether `fld(x,y)*y` is again equal to `x` depends on floating point round-off.

**Comment by simonbyrne** [comment]
This definition still isn't correct, e.g. we have `fld(-1,-3) == 0`, yet `-3*0` is not less than `-1`.

I think we would be better off just defining it as `floor(x/y)` without intermediate rounding, and give an example where it is different.

**Comment by simonbyrne** [comment]
Also, an example of where they're different would be useful.

**Comment by eschnett** [comment]
@simonbyrne In case you replied to me: I suggested  "Largest integer less than or equal to x/y". -1/-3 = 1/3; largest integer less than or equal to that is 0. I think the term "Largest integer less than or equal to" is the same as "rounded towards minus infinity to the next integer", or "floor". If you prefer the term "floor" -- sure, let's go with that.

**Comment by simonbyrne** [comment]
@eschnett No, I was responding to @ranjanan's proposed change. Your suggestion is fine, the main thing is to state that it is calculated without intermediate rounding.

**Comment by Keno** [comment]
:+1: to defining it as `floor(x/y)` without intermediate rounding. Since we now just forward fld to three-argument div, wich is defined as `round(x/y,r)` without intermediate rounding, that would be consistent. In either case, this also pre-dates the .md move, so closing this, but a new PR to improve the docs would be fine.

**Comment by zierenberg** [comment]
Let me start by saying a huge thank you for all the nice development, and sorry for commenting on a closed issue. However, I feel this problem strongly persists - in particular because I just stumbled across this in julia 1.2 (and from the latest version I found this has not changed to the most recent version). 

The problem:
fld(6.0,0.1) = 59 
but 
6/0.1 = 60

I understand that this is due to floating point rounding, but given that the documentation still is defining fld(x::T,y::T) as returning the largest integer less than or equal to x/y this remains confusing. 

I would vote for a word of caution in the documentation with an explicit example.

**Comment by Keno** [comment]
That definition is precise though, 6/0.1 happens to be less than 60 in precise rounding. 

```
julia> 6/big(0.1)
59.99999999999999666933092612453056361837965690217069245739573412231113406246995
```

> I would vote for a word of caution in the documentation with an explicit example.

I think this would be fine. Care to try a new PR with your example that explains the issue?

**Pull request closed without merging**